tests/test_wholebody2.py

```diff
@@ -0,0 +1,227 @@
+        datapoints = fieldmodule.findNodesetByFieldDomainType(Field.DOMAIN_TYPE_DATAPOINTS)
+        self.assertEqual(0, datapoints.getSize())
+
+        # Check coordinates range, volume
+        coordinates = fieldmodule.findFieldByName("coordinates").castFiniteElement()
+        self.assertTrue(coordinates.isValid())
+        minimums, maximums = evaluateFieldNodesetRange(coordinates, nodes)
+        tol = 1.0E-6
+        assertAlmostEqualList(self, minimums, [0.0, -1.976, -0.6113507244855241], tol)
+        assertAlmostEqualList(self, maximums, [8.748998777022969, 1.968, 0.7035640657448223], tol)
+
+        with ChangeManager(fieldmodule):
+            one = fieldmodule.createFieldConstant(1.0)
+            isExterior = fieldmodule.createFieldIsExterior()
+            isExteriorXi3_1 = fieldmodule.createFieldAnd(
+                isExterior, fieldmodule.createFieldIsOnFace(Element.FACE_TYPE_XI3_1))
+            mesh2d = fieldmodule.findMeshByDimension(2)
+            fieldcache = fieldmodule.createFieldcache()
+
+            volumeField = fieldmodule.createFieldMeshIntegral(one, coordinates, mesh3d)
+            volumeField.setNumbersOfPoints(3)
+            result, volume = volumeField.evaluateReal(fieldcache, 1)
+            self.assertEqual(result, RESULT_OK)
+
+            surfaceAreaField = fieldmodule.createFieldMeshIntegral(isExteriorXi3_1, coordinates, mesh2d)
+            surfaceAreaField.setNumbersOfPoints(4)
+            result, surfaceArea = surfaceAreaField.evaluateReal(fieldcache, 1)
+            self.assertEqual(result, RESULT_OK)
+
+            self.assertAlmostEqual(volume, 6.419528010968665, delta=tol)
+            self.assertAlmostEqual(surfaceArea, 35.880031911102506, delta=tol)
+
+        # check some annotationGroups:
+        expectedSizes2d = {
+            "skin epidermis": 308
+            }
+        for name in expectedSizes2d:
+            term = get_body_term(name)
+            annotationGroup = getAnnotationGroupForTerm(annotationGroups, term)
+            size = annotationGroup.getMeshGroup(mesh2d).getSize()
+            self.assertEqual(expectedSizes2d[name], size, name)
+
+            surfaceGroup = AnnotationGroup(region, term)
+            isExterior = fieldmodule.createFieldIsExterior()
+            isExteriorXi3_1 = fieldmodule.createFieldAnd(
+                isExterior, fieldmodule.createFieldIsOnFace(Element.FACE_TYPE_XI3_1))
+            surfaceMeshGroup = surfaceGroup.getMeshGroup(mesh2d)
+            surfaceMeshGroup.addElementsConditional(isExteriorXi3_1)
+
+            surfaceAreaField = fieldmodule.createFieldMeshIntegral(one, coordinates, surfaceMeshGroup)
+            surfaceAreaField.setNumbersOfPoints(4)
+
+            fieldcache = fieldmodule.createFieldcache()
+            result, surfaceArea = surfaceAreaField.evaluateReal(fieldcache, 1)
+            self.assertEqual(result, RESULT_OK)
+            self.assertAlmostEqual(surfaceArea, 35.880031911102506, delta=tol)
```

Comment:
The surface area part needs to be improved. Should not be constructing an AnnotationGroup and filling it -- it exists already in the annotationGroups list, just need to find by name; right now your code is making a group with exterior xi3=1, not checking the actual group..
```
surfaceAnnotationGroup = findAnnotationGroupByName(name)
surfaceMeshGroup = surfaceAnnotationGroup.getMeshGroup(mesh2d)
# then use it.
```
ALSO, should have the answers in the expectedSizes, e.g.:
```
annotationAnswers2d = {
    "skin epidermis": (308, 35.880031911102506)
}
'''
I.e. the RHS of the map is a tuple of size and surface area.

Reply:
Done